toodlext/src/main.rs

```diff
@@ -0,0 +1,355 @@
+    completion_date: Option<i64>,
+
+    labels: Option<Vec<LabelInfo>>,
+}
+
+impl ItemInfo {
+    fn from_item(item: &Item) -> ItemInfo {
+        let due_date = item.due_date.as_ref().map_or(0, to_millis);
+        let completion_date = item.completion_date.as_ref().map_or(0, to_millis);
+        let labels = Some(item.labels.iter().map(LabelInfo::from_label).collect());
+        ItemInfo {
+            uuid: Some(item.uuid.hyphenated().to_string()),
+            name: item.name.clone(),
+            due_date: Some(due_date),
+            completion_date: Some(completion_date),
+            labels,
+        }
+    }
+
+    fn to_item(&self) -> Item {
+        let due_date = self.due_date.map(from_millis);
+        let completion_date = self.completion_date.map(from_millis);
+        let labels = match self.labels {
+            Some(ref labels) => labels.iter().map(|label| label.to_label()).collect(),
+            None => Vec::new(),
+        };
+        Item {
+            id: None,
+            uuid: self.uuid
+                .clone()
+                .and_then(|uuid| Uuid::parse_str(&uuid).ok())
+                .unwrap_or_else(|| Uuid::nil()),
+            name: self.name.clone(),
+            due_date,
+            completion_date,
+            labels,
+        }
+    }
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(tag = "type")]
+enum RequestBody {
+    CreateTodo(ItemInfo),
+    GetTodos,
+    TodoChangeName { uuid: String, name: String },
+    TodoChangeDueDate {
+        uuid: String,
+
+        #[serde(rename = "dueDate")]
+        due_date: i64,
+    },
+    TodoChangeCompletionDate {
+        uuid: String,
+
+        #[serde(rename = "completionDate")]
+        completion_date: i64,
+    },
+    TodoAddLabel { uuid: String, label: LabelInfo },
+    TodoRemoveLabel { uuid: String, name: String },
+    RemoveTodo { uuid: String },
+    AddLabel(LabelInfo),
+    RemoveLabel { name: String },
+    GetLabels,
+}
+
+#[derive(Deserialize, Debug)]
+struct Request {
+    id: i64,
+    body: RequestBody,
+}
+
+impl Request {
+    fn read_from(input: &mut StdinLock) -> Result<Request, Error> {
+        match input.read_u32::<NativeEndian>() {
+            Ok(length) => {
+                let mut message = input.take(length as u64);
+                let mut buffer = Vec::with_capacity(length as usize);
+                eprintln!("Reading request from browser");
+                match message.read_to_end(&mut buffer) {
+                    Ok(_) => {
+                        serde_json::from_slice(&buffer).map_err(|err| {
+                            eprintln!("Error parsing request payload {:?}: {:?}",
+                                      String::from_utf8_lossy(&buffer),
+                                      err);
+                            Error::BadJSON
+                        })
+                    }
+                    _ => Err(Error::IOError),
+                }
+            }
+            _ => Err(Error::IOError),
```

Comment:
You could probably use `map_err` to even better effect here:
```
let length = input.read_u32::<NativeEndian>().map_err(|_err| { Error::IOError })?;
let mut message = input.take(length as u64);
let mut buffer = Vec::with_capacity(length as usize);
eprintln!("Reading request from browser");
message.read_to_end(&mut buffer).map_err(|_err|{ Error::IOError})?;
serde_json::from_slice(&buffer).map_err(|err| {
    eprintln!("Error parsing request payload {:?}: {:?}",
        String::from_utf8_lossy(&buffer), err);
    Error::BadJSON
})

```

Reply:
Nifty!